rts/motoko-rts/src/static_checks.rs

```diff
@@ -0,0 +1,53 @@
+//! Static assertions to make sure object layouts are as expected
+
+use crate::types::*;
+
+use core::mem::{align_of, size_of};
+
+use static_assertions::const_assert_eq;
+
+// TODO: I don't understand why I get a "unused constant" warning for this. Removing it causes
+// compilation failures as expected.
+#[allow(unused)]
+const WORD_SIZE: usize = crate::constants::WORD_SIZE as usize;
+
+// Check platform word size
+const_assert_eq!(size_of::<usize>(), size_of::<u32>());
+const_assert_eq!(size_of::<usize>(), WORD_SIZE);
+
+// Check that sizes of structs are as expected by the compiler
```

Comment:
Where is the compiler expecting these sizes? Maybe extend the comment with a pointer to where these expectations are currently encoded.

Or is this just a sanity check so that we definitely notice changes to these object sizes, but nothing would break if someone changes the size of one of these?

Reply:
Some of these may be implicit in the compiler but for example https://github.com/dfinity/motoko/blob/ede201c60664c3f1ec4e51824ea4e3981b4bd03b/src/codegen/compile.ml#L1446-L1451 expects `Closure`s to be 3 words and https://github.com/dfinity/motoko/blob/ede201c60664c3f1ec4e51824ea4e3981b4bd03b/src/codegen/compile.ml#L2633 expects `Object` header (the `Object` struct in the RTS) to be 3 words.

The idea is if we change the structs in the RTS we'll get a compile error and either update this and the compiler or revert the change. So this checks for accidental changes.

Ideally I think we would have a file that defines object sizes and field offsets, and some kind of preprocessor to generate OCaml and Rust code from that. In GHC we use CPP header files for this. Some other compilers use custom file format and preprocessor.

(I just realized we don't check size of `Object` in here, I'll add it)

Reply:
Ok, maybe a vague
```suggestion
// Check that sizes of structs are as expected by the compiler
// (Expectations are all over the place, e.g. `header_size` definitions in `compile.ml`, calls to `static_closure`, etc.)
```
is helpful.

Reply:
Once we update rustc we can also check that fields offsets are as expected by the compiler, using https://crates.io/crates/memoffset. This crate currently cannot be compiled with the rustc version we're using if we want to use it in `const` context.